# Planning notes

## In attendance

- [x] Mei
- [x] Kannan
- [x] Eric

## Agenda

1. How we are going to perform machine learning model benchmarking.

## Proposed strategy

### Models

- Train patch-GNN model on either homology model or crystal structure data.
- Specific models to try building:
  - Baselines:
    - one-hot + random forest
    - Genentech Model: structural features + random forest (KS has done this; code available)
      - We might not pursue this (because of confidentiality agreement issues)
    - model from [BMC Bioinformatics][bmc]
      - this one is a bit hard-coded
      - [GitHub link][bmc_gh]
    - unirep default weights + random forest
    - unirep default weights + top feed-forward end-to-end model
  - Our models of interest
    - patch-gnn from random weights + top model
    - patch-gnn with weights pre-trained on a small collection of structures

[bmc]: https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-017-1848-9
[bmc_gh]: https://github.com/jcaledo/Prediction_of_Methionine_Oxidation_Sites

### Data cleaning

Key questions:

- What serves as "inclusion criteria" for a sequence?
  - Its full-length sequence is available
  - Its structure is available, or a good quality homology model can be made.

Pitfalls to be aware of:

- UniProt ID might have been updated since dataset publication.
  - Need to resolve these.
- Amino acid residue numbering: write function that checks against raw data file.
- Homology model might omit residues.
- Might need to spot-check homology models.
  - Sanity check using:
    - Ramachandran map (distribution of phi/psi angles in "allowed" regions)
      - This might be automatable. KS will check.
    - Low sequence identity (<20%)
    - Full sequence or a truncated fragment

This way, we should be able to keep the datasets constant between the models.

## Plan going forward

Our data set of choice: Ghesquire 2011.

Basic ideas:

1. Tie loose ends w.r.t. data (upload/download; cleaning)
2. Build baseline models.

Tasks:

- [ ] EM: Work through upload data to datastore (Mei: https://bitbucket.prd.nibr.novartis.net/projects/SDA/repos/sda-rd-data-management/browse?at=refs%2Fheads%2Fdevelop)
- [x] KS: Download UniProt human so that we can pull out proteins of interest.
- Next pair coding session:
  - [ ] MX (driver) (pair code): hack out baseline model code for one-hots + random forest
  - [ ] EM (driver) (pair code): hack out unirep baseline model
  - [ ] KS (driver) (pair code): script out homology model generation
